File: db4oj.tests/src/com/db4o/db4ounit/common/assorted/LazyObjectReferenceTestCase.java
/* Copyright (C) 2007  Versant Inc.  http://www.db4o.com */

package com.db4o.db4ounit.common.assorted;

import com.db4o.*;
import com.db4o.config.*;
import com.db4o.ext.*;
import com.db4o.internal.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.*;


public class LazyObjectReferenceTestCase extends AbstractDb4oTestCase{
	
	public static void main(String[] arguments) {
		new LazyObjectReferenceTestCase().runSolo();
	}
	
	public static class Item {
		
	}
	
	protected void configure(Configuration config) throws Exception {
		super.configure(config);
		config.objectClass(Item.class).generateUUIDs(true);
	}
	
	protected void store() throws Exception {
		for (int i = 0; i < 10; i++) {
			store(new Item());
		}
	}
	
	public void test(){
		Query q = db().query();
		q.constrain(Item.class);
		ObjectSet objectSet = q.execute();
		long[]ids = objectSet.ext().getIDs();
		
		ObjectInfo[] infos = new ObjectInfo[ids.length];
		Item[] items = new Item[ids.length];
		
		for (int i = 0; i < items.length; i++) {
			items[i] = (Item) db().getByID(ids[i]);
			infos[i] = new LazyObjectReference(trans(), (int)ids[i]);
		}
		
		assertInfosAreConsistent(ids, infos);
		
		for (int i = 0; i < items.length; i++) {
			db().purge(items[i]);
		}
		
		db().purge();
		
		assertInfosAreConsistent(ids, infos);
		
	}

	private void assertInfosAreConsistent(long[] ids, ObjectInfo[] infos) {
		for (int i = 0; i < infos.length; i++) {
			ObjectInfo info = db().getObjectInfo(db().getByID(ids[i]));
			Assert.areEqual(info.getInternalID(), infos[i].getInternalID());
			Assert.areEqual(info.getUUID().getLongPart(), infos[i].getUUID().getLongPart());
			Assert.areSame(info.getObject(), infos[i].getObject());
		}
	}
	
	

}

File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Assorted/LazyObjectReferenceTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Query;
using Db4objects.Db4o.Tests.Common.Assorted;

namespace Db4objects.Db4o.Tests.Common.Assorted
{
	public class LazyObjectReferenceTestCase : AbstractDb4oTestCase
	{
		public static void Main(string[] arguments)
		{
			new LazyObjectReferenceTestCase().RunSolo();
		}

		public class Item
		{
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Configure(IConfiguration config)
		{
			base.Configure(config);
			config.ObjectClass(typeof(LazyObjectReferenceTestCase.Item)).GenerateUUIDs(true);
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Store()
		{
			for (int i = 0; i < 10; i++)
			{
				Store(new LazyObjectReferenceTestCase.Item());
			}
		}

		public virtual void Test()
		{
			IQuery q = Db().Query();
			q.Constrain(typeof(LazyObjectReferenceTestCase.Item));
			IObjectSet objectSet = q.Execute();
			long[] ids = objectSet.Ext().GetIDs();
			IObjectInfo[] infos = new IObjectInfo[ids.Length];
			LazyObjectReferenceTestCase.Item[] items = new LazyObjectReferenceTestCase.Item[ids
				.Length];
			for (int i = 0; i < items.Length; i++)
			{
				items[i] = (LazyObjectReferenceTestCase.Item)Db().GetByID(ids[i]);
				infos[i] = new LazyObjectReference(Trans(), (int)ids[i]);
			}
			AssertInfosAreConsistent(ids, infos);
			for (int i = 0; i < items.Length; i++)
			{
				Db().Purge(items[i]);
			}
			Db().Purge();
			AssertInfosAreConsistent(ids, infos);
		}

		private void AssertInfosAreConsistent(long[] ids, IObjectInfo[] infos)
		{
			for (int i = 0; i < infos.Length; i++)
			{
				IObjectInfo info = Db().GetObjectInfo(Db().GetByID(ids[i]));
				Assert.AreEqual(info.GetInternalID(), infos[i].GetInternalID());
				Assert.AreEqual(info.GetUUID().GetLongPart(), infos[i].GetUUID().GetLongPart());
				Assert.AreSame(info.GetObject(), infos[i].GetObject());
			}
		}
	}
}
